rp2040-hal/src/i2c.rs

```diff
@@ -116,22 +121,100 @@ impl SclPin<I2C0> for Gpio21 {}
 impl SdaPin<I2C1> for Gpio26 {}
 impl SclPin<I2C1> for Gpio27 {}
 
-/// I2C peripheral operating in master mode
-pub struct I2C<I2C, Pins> {
+/// Operational mode of the I2C peripheral.
+pub trait I2CMode: Sealed {
+    /// Indicates whether this mode is Controller or Peripheral.
+    const IS_CONTROLLER: bool;
+}
+/// Marker for an I2C peripheral operating as a controller.
+pub enum Controller {}
+impl Sealed for Controller {}
+impl I2CMode for Controller {
+    const IS_CONTROLLER: bool = true;
+}
+/// Marker for an I2C peripheral operating as a peripehral.
+pub enum Peripheral {}
+impl Sealed for Peripheral {}
+impl I2CMode for Peripheral {
+    const IS_CONTROLLER: bool = false;
+}
+
+/// I2C peripheral
+pub struct I2C<I2C, Pins, Mode = Controller> {
     i2c: I2C,
     pins: Pins,
+    mode: PhantomData<Mode>,
 }
 
 const TX_FIFO_SIZE: u8 = 16;
+const RX_FIFO_SIZE: u8 = 16;
 
-fn i2c_reserved_addr(addr: u8) -> bool {
+fn i2c_reserved_addr(addr: u16) -> bool {
```

Comment:
Is the type change here so we can support 10bit addressing in the future?

Reply:
Indeed. For now this is mostly to save us from unnecessary conversion between u8 & u16, since the register takes u16 either way, we may as well just use u16 internally.